### PR TITLE
Introduce Project.clearCaches and Project.duplicate

### PR DESCRIPTION
**Pull request by azinneera**
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/33073

`public void clearCaches();`
`public Project duplicate();`
This will discard the old contexts which includes clearing of compilation caches and diagnostics.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by IMS94** [comment]
@azinneera I tested this implementation locally after applying the pull module code action changes on top of it. Works as expected.

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/directory/BuildProject.java [hunk 1]
```diff
@@ -155,6 +155,13 @@ private BuildProject(ProjectEnvironmentBuilder environmentBuilder, Path projectP
         return Optional.empty();
     }
 
+    @Override
+    public Project duplicate() {
+        BuildProject buildProject = new BuildProject(
```
Comment (sameerajayasoma) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/directory/BuildProject.java:160]: Does this work for situations where I have created the environment with a non-default builder?

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/Project.java [hunk 2]
```diff
@@ -100,6 +101,25 @@ protected void populateCompilerContext() {
         options.put(PROJECT_DIR, this.sourceRoot().toAbsolutePath().toString());
     }
 
+    /**
+     * Refresh the project to clear compilation caches.
+     */
+    public void refresh() {
+        cloneProject(this);
+        CompilerContext compilerContext = this.projectEnvironmentContext()
```
Comment (sameerajayasoma) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/Project.java:109]: Does the`CompilerContext` holds any compilation caches such as langlib etc?

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/directory/BuildProject.java (hunk 1 above)
Comment (azinneera) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/directory/BuildProject.java:160]: No, for the moment we support only with the DefaultEnvironmentBuilder as we don't have use cases for other environments.

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/Project.java (hunk 2 above)
Comment (azinneera) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/Project.java:109]: Yes it does. The caches other than the langlibs are cleared upon refresh. Works similar to edit operations.

**Comment by azinneera** [comment]
@sameerajayasoma As per the offline discussion, renamed the `project.refresh` to `project.clearCaches`

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/directory/BuildProject.java (hunk 1 above)
Comment (sameerajayasoma) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/directory/BuildProject.java:161]: Can we create an issue to capture this limitation? I am pretty sure we are going to encounter this sooner or later.

**Review by sameerajayasoma — APPROVED** [review]
LGTM

**Pull request merged**